// coffeebarx/lib/command.js
// Generated by CoffeeScript 1.6.3
var coffeebar, program;

program = require('commander');

coffeebar = require('./coffeebar');

module.exports.run = function() {
  var options;
  program.version('0.3.0').usage('[options] [path ...]').option('-b, --bare', 'compile without a top-level function wrapper').option('-m, --minify', 'minify output files').option('-M, --map', 'create source maps, appended to the JS file').option('-x, --extmap', 'create source maps, shipped in a separate file').option('-o, --output <path>', 'output path').option('-s, --silent', 'suppress console output').option('-w, --watch', 'watch files for changes');
  program.parse(process.argv);
  options = {
    output: program.output,
    watch: program.watch,
    silent: program.silent || false,
    minify: program.minify,
    sourceMap: program.map,
    extSourceMap: program.extmap,
    bare: program.bare
  };
  if (!options.silent) {
    console.log('');
  }
  coffeebar(program.args, options);
  if (!options.silent) {
    return console.log('');
  }
};
